Add new conditional for resync prefix

gandarez opened the pull request:
This PR adds a new conditional for `resync` prefix. It will allow to resync/update `develop` branch when a hotfix is done at `master`. A new branch must be suffixed with `resync/*`.

- As we've been using the letter `v` as prefix for tags, semver allows to parse tolerant. So I replaced `Parse` by `ParseTolerant`. Less code :)
- Small improvement in the tests by suffixing with `v` all string versions

Closes https://github.com/wakatime/semver-action/issues/8

dron22 commented on README.md:
```diff
@@ -15,6 +15,7 @@ These are the prefixes we expect when `auto` bump:
 - `^feature/.*` - `minor`
 - `^major/.*` - `major`
 - `^misc/.*` - `build`
+- `^resync/.*` - Special case needed to resync base branch into child - Mostly from `master` into `develop`.
```
Can we mention hotfix here, as this is the primary case?

dron22 commented on cmd/generate/generate.go:
```diff
@@ -89,16 +90,14 @@ func Tag(params Params, gc gitClient) (Result, error) {
 	} else {
-		latestTag = prefixRe.ReplaceAllLiteralString(latestTag, "")
-		parsed, err := semver.Parse(latestTag)
+		parsed, err := semver.ParseTolerant(latestTag)
```
:+1:

dron22 commented on cmd/generate/generate_test.go:
```diff
@@ -38,7 +38,7 @@ func TestTag(t *testing.T) {
 		"first non-development tag": {
 			CurrentBranch: "master",
-			LatestTag:     "1.0.0-alpha.1",
+			LatestTag:     "v1.0.0-alpha.1",
```
:+1:

dron22 commented on cmd/generate/generate_test.go:
```diff
@@ -148,9 +148,49 @@ func TestTag(t *testing.T) {
 				IsPrerelease: false,
 			},
 		},
+		"resync into develop when version at master is less than develop": {
```
You write `version at master is less than develop`. Can you explain, how the test data reflects this? As far as I can judge, the test data for the 2 tests look quite similar in this regard. Am I missing something?

dron22 reviewed (commented):
Could you add a test case for hotfix merge into master to `TestTag` in `generate_test.go`? This would help me to understand the whole picture of the hotfix/resync process.

gandarez commented:
@dron22 I changed `resync` method a little bit to accomodate its behaviour in a simpler way.

- Improved `resync` description in README
- Don't need anymore `exclude` pattern in `AncestorTag` since it was conflicting and in some cases dind't return the expected result. So I replaced by branch paramater which is safer.
- Changed a little bit `AncestorTag` mocked func in `generate_test` to return the correct ancestor tag based on branch parameter.

_`greater` and `less` here are used when semver tags are compared one against another._

**_The following scenarios list tags in the creation orders_**. Tags generated from hotfix branches will be always the first ones in the list.

**1st scenario** - If latest tag from semver is greater than ancestor develop (prerelease) - Should generate a new develop tag v1.2.1-alpha.1
- v1.2.1 <- **hotfix here - latest from semver**
- v1.2.0-alpha.1 <- **ancestor at develop**
- V1.1.0-alpha.1
- v1.0.0 <- **ancestor at master**
- v1.0.0-alpha.1

**2nd scenario** - If latest tag from semver is less than ancestor develop (prerelease) - Should bump build version resulting in v1.3.0-alpha.2
- v1.2.1 <- **hotfix here - latest from semver**
- v1.3.0-alpha.1 <- **ancestor at develop**
- v1.2.0-alpha.1
- V1.1.0-alpha.1
- v1.0.0 <- **ancestor at master**
- v1.0.0-alpha.1

dron22 commented:
Question on scenario #1:
- Let's say, resync bumps develop version to `v1.2.1-alpha.1`, like you described.
- Then we merge some doc change into develop and we'll be at `v1.2.1-alpha.2`.
- Then we merge develop into master.
Which tag will be created in master for the new release?

gandarez commented:
> Question on scenario #1:
> 
> * Let's say, resync bumps develop version to `v1.2.1-alpha.1`, like you described.
> * Then we merge some doc change into develop and we'll be at `v1.2.1-alpha.2`.
> * Then we merge develop into master.
> 
> Which tag will be created in master for the new release?

tag will be 💣 lol ... I had the same thought when developing it but I forgot about `doc` and `misc` prefixes. So the only way for it is to always bump patch version. I'll fix it.

gandarez commented:
@dron22 I replaced the new resync method by using the same logic as patch increment, like `bugfix`. By doing in this way we won't have any problems regarding the 1st scenario you mentioned. Thanks 🙏

dron22 reviewed (approved):
Alright, looks good now. Thanks for adjusting :+1:
